### PR TITLE
test(datamodule): add tests for ot module (hungarian matching + collate fns)

### PR DESCRIPTION
**Pull request by ktinubu**
## Summary

- 35 tests covering `src/synth_setter/data/ot.py`: `_hungarian_match` (the OT minibatch-matching used by flow-matching training), the `concatenate`/`stack` torch/numpy dispatch helpers, and the four collate variants (`regular_collate_fn`, `ot_collate_fn` and their tuple/dict inner functions).
- **Flagship adversarial check**: brute-forces every permutation for B≤5 and verifies Hungarian produces the actual minimum-cost pairing. This catches subtle bugs that shape/isinstance tests would miss (e.g. returning `row_ind` instead of `col_ind` for the params permutation).
- Pins other non-obvious behaviors: the inverted noise/params roles (only `params` is permuted; `noise` returns in identity order), co-data alignment in lockstep with `params`, `None` co-data flowing through, and the silent numpy→torch auto-conversion.

No source changes — only the new test file.

## Test plan

- [x] `python -m pytest tests/data/test_ot.py -v` → 35 passed in 0.35s
- [x] `pre-commit run --files tests/data/test_ot.py` → all hooks pass (ruff, pyright, pydoclint, docformatter, interrogate, codespell)
- [ ] CI green on PR

Refs #30

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

- [x] **All 35 tests pass against the real source module** *(Level 1 — promise is end-to-end test discovery + execution against `synth_setter.data.ot`)*
  ```bash
  $ python -m pytest tests/data/test_ot.py -v --no-header 2>&1 | tail -3
  (101 durations < 0.005s hidden.  Use -vv to show these durations.)
  ============================== 35 passed in 0.39s ==============================
  ```

- [x] **Pytest collection picks up exactly 35 tests** *(Level 1 — confirms standard pytest discovery works, not just bespoke invocation)*
  ```bash
  $ python -m pytest --collect-only tests/data/test_ot.py -q --no-header 2>&1 | tail -2
  35 tests collected in 0.12s
  ```

- [x] **Tests catch real regressions (mutation probe on the OT permutation)** *(Level 2 — adversarial; the brute-force oracle is the heart of this PR. If a mutation that breaks OT optimality slips past the tests, the file is just vacuous structure.)*
  Mutated `_hungarian_match` to use `params = params[row_ind]` (identity) instead of `params = params[col_ind]` (the actual OT permutation), re-ran the targeted tests, and confirmed the brute-force/optimality tests fail with the expected numerical mismatches. Source restored after.
  ```bash
  $ # apply mutation, run brute-force + end-to-end OT tests:
  $ python -m pytest tests/data/test_ot.py -k "matches_brute_force or beats_random or end_to_end_with_real" --no-header 2>&1 | tail -6
  FAILED ...test_matches_brute_force_optimum_on_random_inputs[4] - assert 9.551... == 7.054... ± 7.1e-05
  FAILED ...test_matches_brute_force_optimum_on_random_inputs[5] - assert 9.500... == 6.596... ± 6.6e-05
  FAILED ...test_beats_random_pairing_in_expectation - assert 23.4363... < 23.4363...
  FAILED ...test_end_to_end_with_real_hungarian_match_returns_optimal_pairing - assert 7.397... == 7.071... ± 7.1e-05
  ================== 4 failed, 2 passed, 29 deselected in 0.23s ==================
  ```
  Note: B=2/B=3 happen to pass under the chosen seed because identity-vs-OT collide for those particular random draws. B=4 and B=5 catch it. The brute-force oracle is doing its job — exactly the kind of bug a naive `row_ind`/`col_ind` swap would introduce.

- [x] **Tests are safe under pytest-xdist parallel execution** *(Level 2 — adversarial; CI runs `pytest -n auto`. Random seeds + shared `torch.manual_seed` would surface as flakes here.)*
  ```bash
  $ python -m pytest tests/data/test_ot.py -n auto --no-header -q 2>&1 | tail -1
  ============================= 35 passed in 23.99s ==============================
  ```

- [x] **All CI checks green on the PR** *(Level 4 — `gh pr checks 1052`. 11/11 required + optional jobs pass.)*
  ```bash
  $ gh pr checks 1052 | head -11
  Auto-approve status     skipping  0    ...
  Analyze (actions)       pass      55s  ...
  Analyze (python)        pass      57s  ...
  CodeQL                  pass      2s   ...
  check-pr-metadata       pass      6s   ...
  code-quality            pass      49s  ...
  codecov/patch           pass      1s   ...
  run_tests_conda         pass      1m5s ...
  run_tests_macos (macos-latest, 3.10)   pass  2m36s ...
  run_tests_ubuntu (ubuntu-latest, 3.10) pass  2m5s  ...
  run_tests_ubuntu (ubuntu-latest, 3.11) pass  2m3s  ...
  ```

  Mergeability:
  ```bash
  $ gh pr view 1052 --json mergeable,mergeStateStatus -q '{mergeable: .mergeable, mergeStateStatus: .mergeStateStatus}'
  {"mergeStateStatus":"BLOCKED","mergeable":"MERGEABLE"}
  ```
  `BLOCKED` = branch protection rules (review requirement), not a CI/conflict block.

File: tests/data/test_ot.py [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+        torch.manual_seed(1)
+        noise = torch.randn(6, 3)
+        params = torch.randn(6, 3)
+
+        _, matched_params = _hungarian_match(noise.clone(), params.clone())
+
+        # Stack input and output rows, dedupe; if either side had a duplicate or a row
+        # the other side lacked, ``torch.unique`` would return ≠ 6 rows.
+        combined = torch.cat([params, matched_params], dim=0)
+        assert torch.unique(combined, dim=0).shape[0] == 6
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/data/test_ot.py]: Good catch — fixed in 5e82a9a. Replaced the torch.unique row-count proxy with an explicit row-equality matrix bijection check: row_match[i, j] is True iff input row i equals output row j, and we assert sum(dim=0) == 1 and sum(dim=1) == 1 — exactly one match per input row and per output row. This rules out duplicates/drops independent of whether input rows happen to be unique.

File: tests/data/test_ot.py [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+        improves over random pairing for almost every minibatch, and the seeded example confirms
+        the implementation is actually pairing instead of returning identity.
+        """
+        torch.manual_seed(0)
+        noise = torch.randn(8, 4)
+        params = torch.randn(8, 4)
+
+        random_cost = _total_pairwise_cost(noise, params)
+        matched_noise, matched_params = _hungarian_match(noise.clone(), params.clone())
+        ot_cost = _total_pairwise_cost(matched_noise, matched_params)
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/data/test_ot.py]: Good catch — fixed in 5e82a9a. Replaced the RNG-seeded test with test_beats_identity_when_params_is_known_permutation_of_noise, which constructs params as a non-identity row-permutation of noise (so the true OT cost is exactly 0 and the identity-pairing cost is strictly positive). Hungarian must undo the permutation. No torch RNG dependence — the claim is exact, not statistical.
